velox/dwio/common/SelectiveStructColumnReader.cpp

```diff
@@ -287,4 +286,52 @@ void SelectiveStructColumnReaderBase::getValues(
   }
 }
 
+void SelectiveStructColumnReaderBase::setOutputType(
+    const RowTypePtr& outputType) {
+  outputType_ = outputType;
+}
+
+/**
+ * Check the output type against requested type on compatibility.
+ * @param outputType: the output type from user.
+ * @param requestedType: the type from Parquet.
+ */
+void SelectiveStructColumnReaderBase::checkOutputType(
+    const RowTypePtr& outputType,
+    const RowTypePtr& requestedType) {
+  if (outputType == nullptr) {
+    return;
+  }
+  VELOX_CHECK_NOT_NULL(requestedType);
+  for (int i = 0; i < outputType->size(); ++i) {
+    if (!requestedType->containsChild(outputType->nameOf(i)))
+      continue;
+
+    bool isPartitionColumn = false;
+    for (const auto& childSpec : scanSpec_->children()) {
+      if (childSpec->fieldName() == outputType->nameOf(i) &&
+          childSpec->isConstant()) {
+        isPartitionColumn = true;
+        break;
+      }
+    }
+    // Skip the type check for partition column because requested type does not
+    // contain it.
```

Comment:
To fix Spark test: SPARK-18108 Parquet reader fails when data column types conflict with partition ones.